test/rpc_api_mock.go

```diff
@@ -54,20 +55,50 @@ func (mock *mockService) Unpin(ctx context.Context, in *api.Pin, out *struct{})
 }
 
 func (mock *mockService) PinPath(ctx context.Context, in *api.PinPath, out *api.Pin) error {
-	_, err := gopath.ParsePath(in.Path)
+	p, err := gopath.ParsePath(in.Path)
 	if err != nil {
 		return err
 	}
-	*out = *api.PinWithOpts(CidResolved, in.PinOptions)
+
+	var pin *api.Pin
+	if p.IsJustAKey() && !strings.HasPrefix(in.Path, "/ipns") {
+		c, _, err := gopath.SplitAbsPath(p)
+		if err != nil {
+			return err
+		}
+		if c.Equals(ErrorCid) {
+			return ErrBadCid
+		}
+		pin = api.PinWithOpts(c, in.PinOptions)
+	} else {
+		pin = api.PinWithOpts(CidResolved, in.PinOptions)
+	}
+
+	*out = *pin
 	return nil
 }
 
-func (mock *mockService) UnpinPath(ctx context.Context, in string, out *api.Pin) error {
-	_, err := gopath.ParsePath(in)
+func (mock *mockService) UnpinPath(ctx context.Context, in *api.PinPath, out *api.Pin) error {
+	p, err := gopath.ParsePath(in.Path)
 	if err != nil {
 		return err
 	}
-	*out = *api.PinCid(CidResolved)
+
+	var pin *api.Pin
+	if p.IsJustAKey() && !strings.HasPrefix(in.Path, "/ipns") {
+		c, _, err := gopath.SplitAbsPath(p)
+		if err != nil {
+			return err
+		}
+		if c.Equals(ErrorCid) {
+			return ErrBadCid
+		}
+		pin = api.PinWithOpts(c, in.PinOptions)
+	} else {
+		pin = api.PinWithOpts(CidResolved, in.PinOptions)
+	}
+
+	*out = *pin
```

Comment:
This part can be moved to a function and be used in both PinPath() and UnpinPath()